sdk/python/lib/pulumi/output.py

```diff
@@ -101,7 +101,21 @@ def __init__(
         is_known = asyncio.ensure_future(is_known)
         future = asyncio.ensure_future(future)
         # keep track of all created outputs so we can check they resolve
-        SETTINGS.outputs.append(future)
+        with SETTINGS.lock:
+            SETTINGS.outputs.append(future)
+
+        def cleanup(fut: asyncio.Future[T_co]) -> None:
+            if fut.cancelled() or (fut.exception() is not None):
+                # if cancelled or error'd leave it in the deque to pick up at program exit
+                return
+            # else remove it efficiently from the deque
+            outputs = SETTINGS.outputs
+            with SETTINGS.lock:
+                index = outputs.find(fut)
+                outputs[index] = outputs[-1]
+                outputs.pop()
```

Comment:
Why isn't this just `outputs.remove(fut)`? 

I'm not familiar with Python's `deque` so I was playing around with this in a Python 3.10.6 REPL and ran into an error trying to use `find()`:

```
>>> from collections import deque
>>> outputs = deque()
>>> outputs.append(1)
>>> outputs.append(2)
>>> outputs.append(3)
>>> outputs.append(4)
>>> outputs.append(5)
>>> index = outputs.find(3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'collections.deque' object has no attribute 'find'
```